src/associations/has-many.js

```diff
@@ -416,24 +404,35 @@ class HasMany extends Association {
    * Un-associate one or several target rows.
    *
    * @param {Model} sourceInstance instance to un associate instances with
-   * @param {Model|Model[]|string|string[]|number|number[]} [targetInstances] Can be an Instance or its primary key, or a mixed array of instances and primary keys
+   * @param {Model|Model[]|string|string[]|number|number[]} [rawTargetInstances] Can be an Instance or its primary key, or a mixed array of instances and primary keys
    * @param {object} [options] Options passed to `target.update`
    *
    * @returns {Promise}
    */
-  async remove(sourceInstance, targetInstances, options = {}) {
-    const update = {
-      [this.foreignKey]: null,
-    };
-
-    targetInstances = this.toInstanceArray(targetInstances);
+  async remove(sourceInstance, rawTargetInstances, options = {}) {
+    const targetInstances = this.toInstanceArray(rawTargetInstances);
 
     const where = {
-      [this.foreignKey]: sourceInstance.get(this.sourceKey),
-      [this.target.primaryKeyAttribute]: targetInstances.map(targetInstance => targetInstance.get(this.target.primaryKeyAttribute)),
+      [this.target.primaryKeyAttribute]: targetInstances.map(
+        associatedObject => associatedObject[this.target.primaryKeyAttribute],
+      ),
     };
 
-    await this.target.unscoped().update(update, { ...options, where });
+    // update or delete based on the target's foreign key
+    const foreignKeyIsNullable = this.target.rawAttributes[this.foreignKey].allowNull !== false;
+    const target = this.target.unscoped();
+    if (foreignKeyIsNullable && !options.destroy) {
+      const update = {
+        [this.foreignKey]: null,
+      };
+
+      where[this.foreignKey] = sourceInstance.get(this.sourceKey);
+
+      await target.update(update, { ...options, where });
+    } else {
+      const destroyWhere = where[this.target.primaryKeyAttribute];
+      await target.destroy({ ...options, where: destroyWhere });
```

Comment:
both update & destroy should use the same where, they're modifying the same rows

```suggestion
      await target.destroy({ ...options, where });
```

Reply:
@ephys Using this code, I get the error described [here](https://github.com/sequelize/sequelize/pull/14277#issuecomment-1083211443). This has to do with the `Utils.mapWhereFieldName()` call used in `destroy()` but not in `update()` - I was trying to work around that. I looked through the integration tests but couldn't find an example of destroying. Is there an example you could point me to on the correct usage? Sorry about that - I'll keep looking in the meantime.

Reply:
@ephys this passes the entire `where` object - in this test, for example:
```
    [Object: null prototype] {
      id: 1,
      street: 'st1',
      createdAt: '2022-04-09 16:14:22.563 +00:00',
      updatedAt: '2022-04-09 16:14:22.563 +00:00',
      UserId: 1
    }
``` 
to `queryInterface.bulkDelete`.
This throws an invalid value error. 

`queryInterface.bulkUpdate`'s where value takes the form 
```
[Object: null prototype] { id: [ 2 ] }
```
so I think that we need to reformat the `where` object before calling `destroy()`. Do you have any suggestions on how to do this?

Reply:
Nope, I can confirm that the `where` that's used should be exactly the same in both update & destroy branches

If your object looks like this:
```
[Object: null prototype] {
  UserId: 1,
  id: [
    [Object: null prototype] {
      id: 1,
      street: 'st',
      createdAt: '2022-03-30 14:01:23.317 +00:00',
      updatedAt: '2022-03-30 14:01:23.317 +00:00',
      UserId: 1
    }
  ]
}
```

then you have something fishy going on with targetInstances, that's where the issue lies

Reply:
@ephys Gotcha. I've done some more digging - it appears that the object gets formed like that if `this.toInstanceArray(targetInstances)` in `remove()` gets called when the association is defined as such:
```
      User.hasMany(Address, {
        foreignKey: {
          allowNull: false,
        },
      });
```
with `allowNull: false`. Do you know why this would be? I can't find how that'd affect `toInstanceArray`.

Reply:
I'm experiencing the same issue in my Association refactor PR #14280
The issue is that `set()` calls `remove()` with an array of raw entities
That array is then passed to `toInstanceArray`

`toInstanceArray` is supposed to convert these to models, but does it incorrectly because the way it checks if it's already a model is by doing `instanceof this.target` - raw entities will fail this check.

I'm fixing that in my refactor PR

Reply:
Gotcha - okay, mind letting me know when that PR is merged? I don't think I can proceed with tests passing as expected without that call working properly.

Reply:
Sure! I'll ping you when it's ready. I'm going through the last remaining failing unit tests

Reply:
Hey @ephys just wondering if you've gotten a chance to get these to pass?

Reply:
#14280 is ready, but it will take a while before we properly reviewed it. If you want you can temporarily base your branch on the one from that PR to check if things will work once we merge that

Reply:
I don't know if git will be able to handle that gracefully given we're going to squash-merge #14280

But I can take care of copying your changes over once #14280 is merged, I think your PR is basically ready

Reply:
Oh sweet! Yeah that'd be great, thank you. Will be pretty busy for the next few days, but will look to pick up a new issue afterwards.

Reply:
https://github.com/sequelize/sequelize/pull/14280 has been merged!
I'll see about updating this PR when I have a minute :)

Reply:
@ephys I'm clearing up my notifications and trying to ping you in some of those so they're back on your radar